hamilflow/models/pendulum.py

```diff
@@ -0,0 +1,146 @@
+
+
+class PendulumSystem(BaseModel):
+    r"""The params for the pendulum
+
+    :param float omega0: $\omega_0 \coloneqq \sqrt{\frac{U}{I}} > 0$, frequency
+    parameter
+    """
+
+    omega0: float = Field(gt=0)
+
+    def __init__(self, omega0: float, **kwargs: Any) -> None:
+        return super().__init__(omega0=omega0, **kwargs)
+
+
+class PendulumIC(BaseModel):
+    r"""The initial condition for a pendulum
+
+    :param float theta0: $-\frac{\pi}{2} \le \theta_0 \le \frac{\pi}{2}$, the
+    initial angle
+    """
+
+    theta0: float = Field(ge=-math.pi / 2, le=math.pi / 2)
+
+    def __init__(self, theta0: float, **kwargs: Any) -> None:
+        return super().__init__(theta0=theta0, **kwargs)
+
+    @computed_field  # type: ignore[misc]
+    @cached_property
+    def k(self) -> float:
+        r"""A convenient number
+
+        :return float: $\sin\frac{\theta_0}{2}$
+        """
+        return math.sin(self.theta0 / 2)
+
+
+class Pendulum:
+    r"""Generate time series data for a pendulum.
+
+    We describe a generic pendulum system by the Lagrangian action
+    $$
+    S_L\[\theta\] = I \int_{t_0}^{t_1} \mathbb{d}t
+    \left\\{\frac{1}{2} \dot\theta^2 + \omega_0^2 \cos\theta \right\\}\,,
+    $$
+    where $\theta$ is the _angle_ from the vertical to the pendulum;
+    $I$ is the _inertia parameter_ introduced for dimensional reasons,
+    and $\omega_0$ the _frequency parameter_.
+
+    Details are collected in the tutorial.
+    """
+
+    def __init__(
+        self,
+        system: Union[int, float, Dict[str, Union[int, float]]],
+        initial_condition: Union[int, float, Dict[str, Union[int, float]]],
+    ) -> None:
+        if isinstance(system, (float, int)):
+            system = {"omega0": system}
+        if isinstance(initial_condition, (float, int)):
+            initial_condition = {"theta0": initial_condition}
+        self.system = PendulumSystem.model_validate(system)
+        self.initial_condition = PendulumIC.model_validate(initial_condition)
+
+    @cached_property
+    def definition(self) -> Dict[str, float]:
+        """Model params and initial conditions defined as a dictionary."""
+        return dict(
+            system=self.system.model_dump(),
+            initial_condition=self.initial_condition.model_dump(),
+        )
+
+    @property
+    def omega0(self) -> float:
+        return self.system.omega0
+
+    @property
+    def _k(self) -> float:
+        return self.initial_condition.k
+
+    @property
+    def _math_m(self) -> float:
+        return self._k**2
+
+    @computed_field  # type: ignore[misc]
+    @cached_property
+    def freq(self) -> float:
+        r"""Frequency.
+
+        :return float: $\frac{\pi}{2K(k^2)}\omega_0$, where
+        $K(m)$ is [Legendre's complete elliptic integral of the first kind](https://dlmf.nist.gov/19.2#E8)
+        """
+        return math.pi * self.omega0 / (2 * ellipk(self._math_m))
+
+    @computed_field  # type: ignore[misc]
```

Comment:
I am wondering if `@computed_field` actually does things we want it to do outside of `BaseModel`.

Reply:
Not sure what you meant, but I made the `Field`s `frozen`.

Reply:
`@computed_field` is from pydantic and it is used for [serialization etc of `BaseModel`s](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.computed_field). `Pendulum` is not a `BaseModel`.

I had a look at [the code for `computed_field`](https://github.com/pydantic/pydantic/blob/f794f6598bf23895bed6d01242efb47aa50ecff6/pydantic/fields.py#L1037-L1228). Looks like it doesn't do any harm to our code. However, we also don't need to serialize `Pendulum`. In that sense, we don't need this decorator.

Reply:
https://github.com/kausalflow/hamilflow/pull/22/commits/7ca2f09e64ff516e4ada973feb0b63c49422da8d